packages/rest-live-hooks/src/websocket.tsx

```diff
@@ -37,39 +65,79 @@ class WebsocketManager {
             listener.notify.current(update)
           );
         }
-      };
-      this.websocket.onopen = () => {
+      });
+
+      // Closed/Error event handlers set connectionComplete to be a promise
+      // that won't resolve until the connection is reopened to prevent
+      // subscriptions from going through until the connection has been opened.
+      //
+      // NOTE: This is necessary since the subscribe request will fail
+      // since it calls send directly on the socket connection which will
+      // throw an error.
+
+      this.websocket.addEventListener("error", () => {
+        this.isConnectedCallback(false);
+        this.connectionComplete = new Promise<boolean>((resolve) => {
+          this.setConnected = resolve;
+        });
+      });
+
+      this.websocket.addEventListener("close", () => {
+        this.isConnectedCallback(false);
+        this.connectionComplete = new Promise<boolean>((resolve) => {
+          this.setConnected = resolve;
+        });
+      });
+
+      this.websocket.addEventListener("open", () => {
+        // NOTE: This operates under the assumption that no subscription will
+        // arrive before open is done iterating through the listeners, since
+        // buffered messages will be sent twice otherwise
+
+        this.listeners.forEach((listener) => {
+          listener.notify.current({ action: "REVALIDATE" });
+          // if websocket is open then it cannot be null
+          this.websocket!.send(JSON.stringify(listener.request));
+        });
+        this.isConnectedCallback(true);
         this.setConnected(true);
         resolve();
-      };
+      });
     });
   }
-  constructor(url: string) {
+  constructor(url: string, isConnectedCallback: (isConnected: boolean) => any) {
     this.listeners = [];
     this.websocket = null;
     this.url = url;
     this.connectionComplete = new Promise<boolean>((resolve) => {
       this.setConnected = resolve;
     });
+    this.isConnectedCallback = isConnectedCallback;
   }
 
   reset() {
     this.listeners = [];
-    this.websocket.close();
-    this.websocket = null;
+    if (this.websocket) {
+      this.websocket.close();
+      this.websocket = null;
+    }
   }
 
   async subscribe<T extends Identifiable>(
     request: SubscribeRequest<T>,
-    notify: MutableRefObject<(update: ResourceUpdate<T>) => Promise<T[] | T>>,
+    notify: MutableRefObject<
+      (update: ResourceUpdate<T> | RevalidationUpdate) => Promise<T[] | T>
+    >,
     uuid: number
   ) {
     if (this.websocket === null) {
       await this.connect();
```

Comment:
It just occurred to me that this promise will never resolve if the first websocket connection doesn't go through, which we should probably fix.